src/Integration.Vsix/CFamily/FileConfigProvider.cs

```diff
@@ -0,0 +1,116 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using EnvDTE;
+using SonarLint.VisualStudio.Core.CFamily;
+using SonarLint.VisualStudio.Integration.Helpers;
+
+namespace SonarLint.VisualStudio.Integration.Vsix.CFamily
+{
+    internal interface IFileConfigProvider
+    {
+        CFamilyHelper.FileConfig Get(ProjectItem projectItem, string analyzedFilePath, CFamilyAnalyzerOptions analyzerOptions);
+    }
+
+    internal class FileConfigProvider : IFileConfigProvider
+    {
+        private static readonly NoOpLogger noOpLogger = new NoOpLogger();
+        private readonly ILogger logger;
+
+        public FileConfigProvider(ILogger logger)
+        {
+            this.logger = logger;
+        }
+
+
+        public CFamilyHelper.FileConfig Get(ProjectItem projectItem, string analyzedFilePath, CFamilyAnalyzerOptions analyzerOptions)
+        {
+            var analysisLogger = GetAnalysisLogger(analyzerOptions);
+
+            if (!IsFileInSolution(projectItem))
+            {
+                analysisLogger.LogDebug($"[CFamilyHelper] Unable to retrieve the configuration for file '{analyzedFilePath}'. Check the file is part of a project in the current solution.", analyzerOptions);
+                return null;
+            }
+
+            try
+            {
+                // Note: if the C++ tools are not installed then it's likely an exception will be thrown when
+                // the framework tries to JIT-compile the TryGet method (since it won't be able to find the MS.VS.VCProjectEngine
+                // types).
+                return CFamilyHelper.FileConfig.TryGet(analysisLogger, projectItem, analyzedFilePath);
+            }
+            catch (Exception ex) when (!Microsoft.VisualStudio.ErrorHandler.IsCriticalException(ex))
+            {
+                analysisLogger.WriteLine(CFamilyStrings.ERROR_CreatingConfig, analyzedFilePath, ex);
+                return null;
+            }
+        }
+
+        private ILogger GetAnalysisLogger(CFamilyAnalyzerOptions analyzerOptions)
+        {
+            if (analyzerOptions is CFamilyAnalyzerOptions cFamilyAnalyzerOptions &&
+                cFamilyAnalyzerOptions.CreatePreCompiledHeaders)
+            {
+                // In case the requeset is coming from PCH generation, we don't log failures.
+                // This is to avoid redundant messages while navigating unsupported files.
```

Comment:
We don't currently do this for CMake. However, I think we'll only log debug messages saying the compilation database can't be located, so that's probably ok.